### ISSUE TITLE
Support *any* component shortcode + page with renderer plugins

### ISSUE THREAD

**Issue by Holben888** [slinkity/slinkity#55]
**Is your feature request related to a problem? Please describe.**
Our React page and shortcode components are _heavily_ tied to our core Slinkity plugin. This isn't very extensible, since we'd have to write our own bespoke logic for Vue shortcodes + pages, Svelte, Lit Element, etc. Ideally, we could let the community build these using a plugin architecture!

**Describe the solution you'd like**
We'll be basing this on [Astro's UI renderer config](https://docs.astro.build/reference/renderer-reference/), but with _additional_ hooks to set up [components at the page level](https://slinkity.dev/docs/component-pages-layouts/).

We should abstract all logic for setting up component pages and shortcodes to a plugin API. This comes with a few major changes:
- **Our React logic will move to a new package,** with the naming convention `@slinkity/react-renderer`. We'll enforce this for future renderer plugins as well.
- **We'll introduce a `slinkity.config.js` file** for users to apply renderer plugins. We'll likely mirror [Vite's config format](https://vitejs.dev/config), with their [type checking + `defineConfig` helper](https://vitejs.dev/config/#config-intellisense) as a nice-to-have. Here's how you might apply our React renderer:

```js
const reactRenderer = require('@slinkity/react-renderer')

module.exports = defineConfig({
  rendererPlugins: [
    reactRenderer({...rendererConfigOptionsIfAny}),
  ]
})
```

- **We'll drop the requirement to install `react` and `react-dom` from the docs.** Instead, installing the React renderer will import all associated framework libs. Same should go for future renderer plugins!

**Describe alternatives you've considered**
Uhhh manually writing all this renderer logic by hand? Yeah no thanks.

**Comment by esheehan-gsl** [comment]
Pardon my ignorance, but plugins for a plugin seems a little overwrought. If React is so closely tied to the Slinkity core, would it make more sense to instead of think of Slinkity as a family of separate plugins that all add support for different component frameworks? Will there be enough shared functionality between the different component frameworks that there's a benefit to implementing them as plugins of the Slinkity plugin?

I ask because I'm anxiously awaiting support for Svelte, but also wondering if it'd be relatively easy to set up support for Svelte using a custom shortcode and the new `eleventyConfig.addExtension`.

**Comment by Holben888** [comment]
Fair question @esheehan-gsl! So to clarify, I wouldn't consider Slinkity to be _just_ a family of independent plugins. It's mostly an interface / glue to connect the Vite bundler to your existing 11ty projects. Shortcodes and component pages are honestly a side effect of this integration, with many possibilities beyond that.

If you were to write your own Svelte shortcode or `addExtension` page today using "plain" 11ty, you'd hit some pretty significant roadblocks:
- how do you access the Vite bundler to process that component in development and production? 
- how will hot module reloading (HMR) work?
- how can you hydrate that component at the right time using [hydration modes](https://slinkity.dev/docs/partial-hydration/)? And how can you opt-out of clientside hydration altogether?

We think "custom renderers" would let us answer that question for you. We'll handle all the logistics of supporting partial hydration modes, and pass off our Vite server instance so you can support your framework of choice.

The Astro project has chosen a renderer plugin model to address this issue, allowing them to quickly support new frameworks as they arise. Heck, the maintainer of Solid managed to add support in a matter of minutes! There's honestly no way I can manually support every component framework under the sun (React and Svelte are all I know at the moment). So we're working closely with the Astro team to understand how their custom renderer API works, so we can then bring a similar experience to our project.

Agreed that plugins-within-plugins are a bit extreme though. We'll likely just call them "custom renderers" instead of "renderer plugins" to reduce confusion 😄 

Hope that makes sense! I'm actually recording a talk for Svelte Summit, so I'll at least have a Svelte canary branch ready within the next week 😉

**Comment by esheehan-gsl** [comment]
I'm very excited about the Svelte support—even in canary. :grin:

I hear what you're saying about getting everything wired up correctly. Partially what I was wondering was how much of setting up Eleventy with Vite and getting HMR going (I haven't tried yet) is just a matter of setting your NPM scripts up the right way, and thus maybe more of an Eleventy + Vite starter, then an Eleventy plugin.

I guess, pragmatically, the only thing I'd be concerned about is that setting up Slinkity starts to feel like configuring Babel after Babel went to that micro-library architecture where you could no longer just `npm install babel`, but you had to install the CLI and all these different plugins just to get it to do anything.

Regardless, I'm excited to see how this shakes out.

**Comment by Holben888** [comment]
@esheehan-gsl I totally understand where you're coming from. When I started this project it really _was_ just running the eleventy and Vite processes side-by-side using concurrently! But I quickly discovered the nuance of _prerendering_ those components at buildtime as well. We needed to set up our own middleware to server render (or build-time render I suppose?) using [Vite SSR](https://vitejs.dev/guide/ssr.html). This meant managing the server ourselves instead of using Vite's built-in server. 

Now I will say: **our end goal is the eliminate the slinkity CLI tool entirely, and spin up Vite within an 11ty plugin.** We're just blocked on 
1. Being able to disable 11ty's Browsersync reload events and use Vite's instead
2. Using asynchronous code in our config. Speaking with Zach (core 11ty maintainer), this will take more work than you might think. [Here's a WIP PR](https://github.com/11ty/eleventy/pull/1429/files) that'll likely need to get rewritten

So I'm with you! We're just in this awkward half-meta-framework state until 11ty can support our ambitions.
